lib/core/src/Cardano/Wallet/Primitive/Types.hs

```diff
@@ -491,17 +499,45 @@ instance NFData Address
 instance Buildable Address where
     build = build . toText
 
+instance ToText Address where
+    toText = T.decodeUtf8
+        . convertToBase Base16
+        . getAddress
+
 instance FromText Address where
-    fromText x = maybe
-        (Left $ TextDecodingError err)
-        (pure . Address)
-        (decodeBase58 bitcoinAlphabet $ T.encodeUtf8 x)
+    fromText = bimap textDecodingError Address
+        . convertFromBase Base16
+        . T.encodeUtf8
       where
-        err = "Unable to decode Address: expected Base58 encoding"
+        textDecodingError = TextDecodingError . show
 
-instance ToText Address where
-    toText = T.decodeUtf8 . encodeBase58 bitcoinAlphabet . getAddress
+-- | An abstract class to allow encoding of addresses depending on the target
+-- backend used.
+class EncodeAddress t where
```

Comment:
this is probably very stupid question but I am confused which instance of `EncodeAddress` and `DecodeAddress` are picked here?

I don't see any instance/implementation of these two classes provided (except `DummyTarget` in spec)?

Reply:
For example  - which impl will be picked in L449 https://github.com/input-output-hk/cardano-wallet/pull/313/files#diff-e1113f15531f8432f00f243428125671R449 ?

Reply:
That's the point. The instance is required when we create a server (see: [Api/Server](https://github.com/input-output-hk/cardano-wallet/blob/KtorZ/239/address-format-in-shelley-era/lib/core/src/Cardano/Wallet/Api/Server.hs#L127-L142) and we happen to make a choice at the call-site in [exe/wallet/Main.hs](https://github.com/input-output-hk/cardano-wallet/blob/KtorZ/239/address-format-in-shelley-era/exe/wallet/Main.hs#L336-L337)

Instances for these various parametric interfaces are defined in the `Compatbility` modules of each backend option, for instance for `http-bridge`:

https://github.com/input-output-hk/cardano-wallet/blob/KtorZ/239/address-format-in-shelley-era/lib/http-bridge/src/Cardano/Wallet/Compatibility/HttpBridge.hs 

(note that I haven't yet defined the instance for `EncodeAddress` and `DecodeAddress` here, but that'd basically be where we pick `toBase58` and `fromBase58`).

We would have similar instances in `jormungandr` picking `toBech32` and `fromBech32`.

Reply:
make sense now. Thanks for detailed walkthrough